src/core/ledger/grainIntegration.js

```diff
@@ -0,0 +1,124 @@
+  processDistributions: boolean,
+|};
+
+/**
+ * This function definition is implemented by Grain Integrations.
+ * Grain integrations allow distributions to be executed programmatically
+ * beyond the ledger. However, an integration might have some side-effects
+ * that require the ledger to be updated, and it therefore has the option of
+ * returning a list of of ledger operations. The ledger will update the ledger
+ * if accounting is enabled. Otherwise, grain balances will be tracked
+ * elsewhere.
+ */
+export type grainIntegration = (
+  AccountDistributions,
+  Currency
+) => ?IntegrationResult;
+
+///////////////////
+// Helper functions
+///////////////////
+
+// TODO @topocount: Refactor interface using instance/config properties
+// after grainConfig is updated to include The payout currency details and
+// sink identity
+export function executeGrainIntegration(
+  ledger: Ledger,
+  integration: grainIntegration,
+  distribution: Distribution,
+  currency: Currency,
+  accountingEnabled: boolean,
+  processDistributions: boolean,
+  sink?: IdentityId
+): Ledger {
+  const {distributions, payoutAddressToId} = buildDistributionIndexes(
+    ledger,
+    distribution,
+    JSON.stringify(currency)
+  );
+  // Need to receive actual allocations so users don't lose funds if
+  // decimals are truncated in L2 or in some other environment that must modify
+  // the fixed-point amount for some reason.
+  let result;
+  try {
+    result = integration(distributions, currency);
+    if (processDistributions) ledger.runIntegration(distribution.id);
+  } catch (e) {
+    throw new Error(`Grain Integration failed: ${e}`);
+  }
+  if (result && sink && accountingEnabled && processDistributions) {
+    const {transferredGrain} = result;
+    for (const [address, {amount, memo}] of transferredGrain.entries()) {
+      const recipientId = payoutAddressToId.get(address);
+      if (!recipientId)
+        throw new Error(`Invalid recipient address: ${address}`);
+      ledger.transferGrain({
+        from: recipientId,
+        to: sink,
+        amount,
+        memo: `Integrated Distribution: ${memo}`,
+      });
+    }
+  }
+  return ledger;
+}
+
+export function buildDistributionIndexes(
+  ledger: Ledger,
+  distribution: Distribution,
+  currencyId: CurrencyId
+): {distributions: AccountDistributions, payoutAddressToId: PayoutAddressToId} {
+  const distributions = new Map();
+  const payoutAddressToId = new Map();
+  const balances = getDistributionBalances(distribution);
+  for (const [id, amount] of balances.entries()) {
+    const {payoutAddresses, identity} = ledger.account(id);
+    const address = payoutAddresses.get(currencyId);
+    if (!address) continue;
+    // need to allow for identities that have since been merged to still claim
+    // funds if accounts are merged between a grain distribution a
+    // grainIntegration call.
+    const total = distributions.get(address) ?? G.ZERO;
```

Comment:
I think this is the first time we've used ?? which is an ES2020 operator. Just wondering what the compatibility implications of this are.